include/deal.II/base/aligned_vector.h

```diff
@@ -1036,6 +1107,302 @@ AlignedVector<T>::fill(const T &value)
+    Utilities::MPI::this_mpi_process(shmem_group_communicator) ==
+    Utilities::MPI::n_mpi_processes(shmem_group_communicator) - 1;
+
+  // **** Step 2 ****
+  // We then have to send the state of the current object from the
+  // root process to one exemplar in each shmem group. To this end,
+  // we create another subcommunicator that includes the ranks zero
+  // of all shmem groups, and because of the trick above, we know
+  // that this also includes the original root process.
+  //
+  // There are different ways of creating a "shmem_roots_communicator".
+  // The conceptually easiest way is to create an MPI_Group that only
+  // includes the shmem roots and then create a communicator from this
+  // by way of MPI_Comm_create or MPI_Comm_create_group. The problem
+  // with this is that we would have to exchange among all processes
+  // which ones are shmem roots and which are not. This is awkward.
+  //
+  // A simpler way is to use MPI_Comm_split that uses "colors" to
+  // indicate which sub-communicator each process wants to be in.
+  // We use color=0 to indicate the group of shmem roots, and color=1
+  // for all other processes -- the latter will simply not ever do
+  // anything among themselves with the communicator so created.
+  //
+  // Using MPI_Comm_split has the additional benefit that, just as above,
+  // we can choose where each rank will end up in the shmem roots communicator.
+  // We would again like to set key=0 for the original root_process -- in other
+  // word, we would then know that among the shmem roots, the original root has
+  // rank=0. But, just like above, this doesn't appear to be working, so we put
+  // the origin root at the *end* of the shmem roots communicator.
+  //
+  // In any case, this makes it easy to next determine which process among the
+  // shmem roots is the one who initiates the broad cast operation mentioned
+  // above
+  MPI_Comm shmem_roots_communicator;
+  {
+    const int key =
+      (Utilities::MPI::this_mpi_process(communicator) == root_process ?
+         Utilities::MPI::n_mpi_processes(communicator) :
+         0);
+
+    const int ierr = MPI_Comm_split(communicator,
+                                    /*color=*/
+                                    (is_shmem_root ? 0 : 1),
+                                    key,
+                                    &shmem_roots_communicator);
+    AssertThrowMPI(ierr);
+
+    // Again verify the explanation from above
+    if (Utilities::MPI::this_mpi_process(communicator) == root_process)
+      Assert(Utilities::MPI::this_mpi_process(shmem_roots_communicator) ==
+               Utilities::MPI::n_mpi_processes(shmem_roots_communicator) - 1,
+             ExcInternalError());
+  }
+
+  // Now let the original root_process broadcast the current object to all
+  // shmem roots. We know that the last rank is the original root process that
+  // has all of the data
+  if (is_shmem_root)
+    Utilities::MPI::broadcast(
```

Comment:
doesn't this need a ``*this = Utilities::MPI::broadcast(...`` ?

Reply:
Oh, you're totally right. This only ever worked because I was on a single shmem domain and the shmem_communicator has only one process! Good find!